core/headerchain.go

```diff
@@ -122,87 +122,121 @@ func (hc *HeaderChain) GetBlockNumber(hash common.Hash) *uint64 {
 	return number
 }
 
-// WriteHeader writes a header into the local chain, given that its parent is
-// already known. If the total difficulty of the newly inserted header becomes
-// greater than the current known TD, the canonical chain is re-routed.
+// WriteHeaders writes a chain of headers into the local chain, given that the parents
+// are already known. If the total difficulty of the newly inserted chain becomes
+// greater than the current known TD, the canonical chain is reorged.
 //
 // Note: This method is not concurrent-safe with inserting blocks simultaneously
 // into the chain, as side effects caused by reorganisations cannot be emulated
 // without the real blocks. Hence, writing headers directly should only be done
 // in two scenarios: pure-header mode of operation (light clients), or properly
 // separated header/block phases (non-archive clients).
-func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, err error) {
-	// Cache some values to prevent constant recalculation
+func (hc *HeaderChain) WriteHeaders(headers []*types.Header, pwCallbackFn PWCallback) (ignored, imported int, status WriteStatus, lastHash common.Hash, lastHeader *types.Header, err error) {
+	if len(headers) == 0 {
+		return ignored, imported, NonStatTy, lastHash, nil, err
+	}
+	ptd := hc.GetTd(headers[0].ParentHash, headers[0].Number.Uint64()-1)
+	if ptd == nil {
+		return ignored, imported, NonStatTy, lastHash, nil, consensus.ErrUnknownAncestor
+	}
+	type numberHash struct {
+		number uint64
+		hash   common.Hash
+	}
+	lastHash = headers[0].ParentHash // Already validated above
 	var (
-		hash   = header.Hash()
-		number = header.Number.Uint64()
+		lastNumber = headers[0].Number.Uint64() - 1 // Last successfully imported number
+		externTd   *big.Int                         // TD of successfully imported chain
+		inserted   []numberHash                     // Quick lookup of number/hash for the chain
```

Comment:
I was wondering if you could get away with using `headers` instead of `inserted` since you will know `lastNumber` and have the hash and number of the first header at `header[0]`. I'm not a big fan of method-scoped `struct` definitions personally.

Reply:
One problem with that is that I don't want to write the hashes until later, and if I don't stash the hashes here, I'll have to call `Hash()` on each header again. 
`types.Block.Hash()` calls `Header.Hash` and remembers the hash internally, but headers don't store it internally.

Reply:
> but headers don't store it internally.

Perhaps this is something that could be done (and could potentially benefit the rest of the code base) (maybe in the future).